### DISCUSSION TITLE
How to structure an app using GraphQL and Firebase Auth.

### DISCUSSION THREAD

**Discussion by JulioBarros** [vercel/next.js#11307] [Help]
Hi,

I'm new to NextJS and am enjoying it so far but am finding it challenging to wrap my head around how to structure an app that has some static pages, regular SSR pages with GraphQL and uses GraphQL with Firebase/JWT auth for some of the dynamic pages (clients side rendering).

In particular, firebase needs to be initialized when someone logs in and I need to maintain that initialized app and JWT token across pages so that pages can render in anonymous or logged in mode. Ideally pages would be SSR for anonymous users and client side / dynamic for logged in users so they could edit and view changes right away.

In regular CRA apps I could use a context to track the auth state ... but with NextJS I seem to lose the context and user auth info when I don't expect to. I'm guessing because some pages end up being SSR ... but being new I don't have a good enough mental model to figure out how to address this and/or if I'll be able to do what I want/need with NextJS.

I've looked at many of the examples and they are useful in isolation but don't really seem to address this particular problem/configuration and often don't explain their approaches well enough for me to grok.

Does anyone know of a good tutorial, article or documentation link that I could review for insight?

Also, is there a discord, slack, gitter, etc. for NextJS? This discussion feature is all I've found but it leaves a lot to be desired like search.

Thanks in advance.

Julio

**Comment by Penny-Cho** [comment]
Same here. tried to combine apolloClient with Firebase auth using with-apollo HOC got failed. 
I get tokens only after gql query has been sent to the server.

I have tried these ways
1. sending tokens by creating apolloClient  on _app.js and wrapping with apollo provider got failed and was getting "client.watchquery is not a function" errors.
2. sending tokens via with-apollo HOC got failed because it does not sending the tokens to the server when the query was executed. functions like Firebase's auth().currrentUser  / firebase.auth().onauthstatechanged  could return tokens only after app was rendering more than 2 times.
3. sending headers using useQuery(QUERIES, { context:{headers: {tokens: tokens}} }  directly on individual pages got failed as same as the issues in no.2 way.

also tried getintialprops, getdatafromtrees... none of these worked.

wish there is some good solutions about these issues.

**Comment by codeGTI** [comment]
Same issue here, did you find any solutions for it?
